Simplify CMakeLists

Flamefire opened the pull request:
This is based on #15 

Some notes:

-  `cmake_minimum_required(VERSION 3.5)` has to be before `project` (it sets policies)
- `otf2xx_VERSION` is not required. CMake already has the generic `PROJECT_VERSION`
- `otf2xx_allowed_duration_types` introduced to avoid redundant mentioning of the values
- CMake variable `BUILD_SHARED_LIBS` is used to avoid duplicate code when adding libraries
- `EXCLUDE_FROM_ALL` is removed. It can be added by consumers in `add_subdirectory` (see https://github.com/tud-zih-energy/otf2xx/pull/15#discussion_r205072705) 

About the "code bloat" for the default values:

As discussed on the call I tested the order of `set(` and `set(...CACHE`). The CMake docu states, that the last `set` value is the valid one and a `set(...CACHE` does not set the cache variable when it exists. Hence this shorter code **does** change behavior as e.g. a `set(OTF2XX_CHRONO_DURATION_TYPE "nanoseconds")` before including this file is ignored. However in projects including this with `add_subdirectory` (only relevant use case) one can use `set(OTF2XX_CHRONO_DURATION_TYPE "nanoseconds" CACHE STRING "")` instead which I would prefer to avoid cluttering this CMakeLists.   
However I'm willing to change this if wanted, but please think about it.

Note that the current implementation is flawed:

> if(NOT OTF2XX_USE_STATIC_LIBS)
>     set(OTF2XX_USE_STATIC_LIBS ON)
> endif()

This will ALWAYS set the variable to `ON`. So at least for that the behavior would not change (except being changeable at all now) with the proposed change.

bmario commented on CMakeLists.txt:
```diff
@@ -29,41 +29,35 @@
 cmake_minimum_required(VERSION 3.5)
 
-set(otf2xx_VERSION 1.0)
+project(otf2xx VERSION 1.0.0)
```
> The VERSION option may not be used unless policy CMP0048 is set to NEW.

Flamefire replied:
That is the case for CMake >= 3.0:

- https://devdocs.io/cmake~3.5/policy/cmp0048#policy:CMP0048 (Introduced in 3.0)
- https://devdocs.io/cmake~3.5/command/cmake_policy (All policies introduced in the specified version or earlier will be set to use NEW behavior)
- https://devdocs.io/cmake~3.5/command/cmake_minimum_required (command implicitly invokes cmake_policy)

bmario replied:
I didn't know that cmake_minim_required also sets policies to NEW. I tried it in the past and got the unpleasant error:
```
CMake Error at CMakeLists.txt:2 (project):
  VERSION not allowed unless CMP0048 is set to NEW
```

I probably had `project` and `cmake_minimum_required` in the wrong order :(

Flamefire replied:
Yes. They state that in the docu. In this project it was the wrong way round (also fixed here)

Flamefire commented:
I rebased this and added the required calls for finding the dependencies (Boost, MPI, add MPI-target for CMake < 3.9 [please don't raise the requirement just for that])

Summary of breaking changes and impact:
- EXCLUDE_FROM_ALL removed: Should be added to `add_subdirectory` instead, impact extremly low (possible extra build of 2 cpp files if reader is not used)
- `OTF2XX_*` vars react only to cache values
  - `OTF2XX_USE_STATIC_LIBS` did not work before, now it does
  - `OTF2XX_CHRONO_DURATION_TYPE`: E.g. in trace-merger the default was used anyway
  - `OTF2XX_WITH_MPI`: Adding `CACHE BOOL ""` to the set command makes it work again

Any more changes required?

bmario commented on CMakeLists.txt:
```diff
@@ -87,15 +81,22 @@ if(OTF2XX_WITH_MPI)
     )
+    find_package(Boost REQUIRED COMPONENTS mpi)
+    find_package(MPI REQUIRED COMPONENTS CXX)
+    if(NOT TARGET MPI::MPI_CXX) # For CMake < 3.9
```
This is pure madness. I refuse to merge code duplicates based on some random CMake version decision.
Either a) without these checks, because they are not required, or b) with a CMake version, which provides the targets.

Flamefire replied:
Disclaimer: I want to use OTF2xx in a project where the requirement is CMake 3.8 and raising that is/might be problematic.   
CMake 3.8 was chosen (e.g.) to support proper C++ standards which is very important (the other solutions have various shortcomings not easily worked around)

What is wrong with this? When sticking to CMake 3.8 this **is** required. The alternative would be to add a bunch of `target_include_directories`, `target_compile_definitions` and `target_link_libraries` to the code. But this is forward compatible so it allows improvements from using newer CMake versions without any changes.

So what do you mean by "code duplicates"? Except the outer if-endif the alternative would be the same (and IMO even more verbose)

bmario replied:
The code duplicate is to define MPI targets, just because someone decided CMake newer than some random number is the work of the devils. I don't like that and it either stays like it is, because It is not required. The contract is: "If you want to have MPI (by setting OTF2XX_WITH_MPI to ON), then you have to provide the dependencies yourself.".

Or it will be merge with MPI targets from a decently new CMake version.

Flamefire replied:
>  because someone decided CMake newer than some random number is the work of the devils

Did you ever had to work in a project where you did not have the privilege to dictate the requirements? There might be reasons why a specific version was chosen so assuming it is a "random number" is mislead. And especially a library meant to be used by others should aim to keep the requirements low so it can be used by as many targets as possible.

Code duplication would be if every consumer using CMake 3.8 would be forced to define the MPI:MPI_CXX target himself.

> The contract is: "If you want to have MPI (by setting OTF2XX_WITH_MPI to ON), then you have to provide the dependencies yourself.".

And how would you handle tests of OTF2xx itself? Libraries should be self-contained and currently a simple `cmake .. -DOTF2XX_WITH_MPI=ON` of this project would fail.

I don't see anything wrong with the standard, intended and straight forward use of the CMake 3.8 findMPI module. However as a compromise I would change the code inside the if-endif to a `message(FATAL_ERROR` mentioning the need of CMake >=3.9 or the MPI::MPI_CXX target.

bmario replied:
We are talking about (Boost)MPI. I see no scenario, where you can use MPI "self-contained" in a library, especially when the interface demands a valid boost communicator being passed to it. You therefore have to setup a suitable communicator and provide it to otf2xx and therefore have to do the dependency checks yourself anyway.

Flamefire replied:
1. Being able to do `cmake .. -DOTF2XX_WITH_MPI=ON` could be used in CI to check that all MPI code is valid too
2. Enabling MPI does not mean that it must be used. Correct me if I'm wrong, but you can use all the serial stuff from otf2xx even with `-DOTF2XX_WITH_MPI=ON`.
3. You could add an MPI test to otf2xx tests where the required `MPI_Init` (or whatever) is called and the MPI code is tested in the same build as the non-MPI tests.

This is what I mean by "self-contained".

bmario replied:
In principle, I'm not opposed to add and expose the MPI dependency. I'm opposed against duplicating code, which exist in CMake 3.9 onwards. I won't merge it this way. Period.

If your project insist on using CMake 3.8, fine, but don't flame me for not wanting to maintain manually added backwards compatibility CMake code for a theoretical existing feature nobody asked for.

Flamefire replied:
I don't see where I'm "flaming you", but rather the opposite: "pure madness", "random ... decision", "work of the devils", "I don't like that", ... mixed with claims for which I brought arguments that they are wrong ("(Boost)MPI is not required")   
I brought arguments for why the MPI dependency should be included in the libraries CMake and why it is beneficial to use a method, that works in more versions of CMake.   
I provided code that achieves this. It is easy, short, understandable and (pretty much) the standard way for CMake modules, not defining targets. I don't see where this hurts anyone: No need to change this, no need to write it (as I already did).

It seems we have different goals. I want this library to be as compatible as reasonably possible so it can be widely used. This means keeping requirements low. And if one can avoid raising a requirement by ~5 lines of trivial code, I deem that reasonable. The alternative is less people being able to use this. And if one goes down that path, where do you stop? "We can save 2 lines, when using C++20", "Another 3 when using CMake 3.12", "Oh hey, Boost 1.67 has this cool thing!", ... I just ask: Is this worth it? What does it cost you to have those 5 lines?

However I already offered to replace this by a simple error message. Currently this CMake requires 3.8 but uses a 3.9 feature and nowhere mentions this. So either:
- Raise the minimum (which I'd ask to not, as e.g. we won't be able to use this)
- Copy in the FindMPI from a newer CMake (ew...)
- Use this simple code to allow compatibility with 3.8+
- Add a message or at least docu

Flamefire commented:
As requested the definition of `MPI::MPI_CXX` for CMake < 3.9 was removed. An error will be raised instead so a user doesn't just get a linker error much later.

bmario commented on CMakeLists.txt:
```diff
@@ -143,7 +137,7 @@ if(CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR)
     include(CMakePackageConfigHelpers)
     write_basic_package_version_File("otf2xxConfigVersion.cmake"
-        VERSION ${otf2xx_VERSION}
+        VERSION ${PROJECT_VERSION}
```
Is there a compelling reason to use ${PROJECT_...} variables in general? It's neither more portable, because the same statement contains `otf2xx` anyways nor is it shorter.

Flamefire replied:
There is probably a good guideline for that but what I can think of:
- Shorter as no `set(otf2xx_VERSION 1.0)` required, instead added to `project` command
- Standard names, so less confusion (OTF2XX_VERSION or otf2xx_VERSION?) and better portability (e.g. in `write_basic_package_version_File` you just need to change the name once)
- more version variables for free: `PROJECT_VERSION, <PROJECT-NAME>_VERSION, PROJECT_VERSION_MAJOR, `<PROJECT-NAME>_VERSION_MAJOR,...`

So just minor reasons, but IMO the `PROJECT_` variables are better for reusing CMake code. Note that you can e.g. use `write_basic_package_version_File("${PROJECT_NAME}ConfigVersion.cmake" ...`

Final note: Adding the version to `project(otf2xx` would also allow using `${otf2xx_VERSION}`.

bmario replied:
> Final note: Adding the version to project(otf2xx would also allow using ${otf2xx_VERSION}.

That's why I'm asking...

Flamefire replied:
Ok didn't got that from the question.
I prefer the `PROJECT_*` variables (including `PROJECT_NAME`, `PROJECT_BINARY/SOURCE_DIR`...), as it makes the code more generic. One could put the packaging code into an own function and reuse it. I'm surprised `CMakePackageConfigHelpers` does not do that.

bmario commented on CMakeLists.txt:
```diff
@@ -87,15 +81,15 @@ if(OTF2XX_WITH_MPI)
 endif()
 
-set(READER_SRCS src/reader/callback/definitions.cpp src/reader/callback/events.cpp)
-if(OTF2XX_USE_STATIC_LIBS)
```
When I'm thinking about it, it's good that the shared build never worked, as the shared library will not be installed.

Flamefire replied:
Why? You install the target and if the target is a shared lib it will be installed. Or am I missing anything?

bmario replied:
If you have otf2xx as a git submodule using add_subdirectory(), then there is no install().

Flamefire replied:
Yes. But in that case the consumer is responsible for making sure the library exists (however he does that), when building and installing OTF2xx everything is installed correctly which is IMO the only important thing here.
